// react-native-dynamic-style-processor/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.process = process;

var _reactNative = require("react-native");

var _cssViewportUnitsTransform = require("css-viewport-units-transform");

var _reactNativeCssMediaQueryProcessor = require("react-native-css-media-query-processor");

var _microMemoize = require("micro-memoize");

var _microMemoize2 = _interopRequireDefault(_microMemoize);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function omit(obj, omitKey) {
  return Object.keys(obj).reduce(function (result, key) {
    if (key !== omitKey) {
      result[key] = obj[key];
    }
    return result;
  }, {});
}

var omitMemoized = (0, _microMemoize2.default)(omit);

function viewportUnitsTransform(obj, matchObject) {
  var hasViewportUnits = "__viewportUnits" in obj;

  if (!hasViewportUnits) {
    return obj;
  }
  return (0, _cssViewportUnitsTransform.transform)(omitMemoized(obj, "__viewportUnits"), matchObject);
}

function mediaQueriesTransform(obj, matchObject) {
  var hasParsedMQs = "__mediaQueries" in obj;

  if (!hasParsedMQs) {
    return obj;
  }
  return (0, _reactNativeCssMediaQueryProcessor.process)(obj, matchObject);
}

function process(obj) {
  var matchObject = getMatchObject();
  return viewportUnitsTransform(mediaQueriesTransform(obj, matchObject), matchObject);
}

function getMatchObject() {
  var win = _reactNative.Dimensions.get("window");
  return {
    width: win.width,
    height: win.height,
    orientation: win.width > win.height ? "landscape" : "portrait",
    "aspect-ratio": win.width / win.height,
    type: "screen"
  };
}
